packages/neuron-wallet/tests/services/tx/transaction-service.test.ts

```diff
@@ -0,0 +1,243 @@
+import os from 'os'
+import fs from 'fs'
+import path from 'path'
+import TransactionService, { SearchType } from '../../../src/services/tx/transaction-service'
+import { TransactionStatus } from '../../../src/models/chain/transaction'
+import { saveTransactions, loadTransactions, initConnection, closeConnection, saveAccounts } from '../../setupAndTeardown'
+
+const transactions = loadTransactions()
+
+describe('Test transaction service', () => {
+  beforeEach(async () => {
+    await initConnection()
+    await saveAccounts()
+    return saveTransactions(transactions)
+  })
+
+  afterEach(() => {
+    return closeConnection()
+  })
+
+  describe('Test filter search type', () => {
+    it('Should return empty', () => {
+      const actual = TransactionService.filterSearchType('')
+      expect(actual).toBe(SearchType.Empty)
+    })
+
+    it('Should return address', () => {
```

Comment:
Most of these tests are having both `condition (when)` and `then` located inside a `it()`.
For better readability and reusability of the preconditions for the different tests, how about we organize the tests in a bit more expressive way, such as 
```
describe('when', () => {
  beforeEach(() => {
    //precondition logics
  })
  it('then', () => {
    //assert logics
  })
})
```

Reply:
Updated in https://github.com/nervosnetwork/neuron/pull/1733/commits/00d73419d36f406d3eed6753679f1de1b07055db

Reply:
Maybe I miss something. The `when` and `then` are still within a `it()` in this new commit.

Reply:
Do you mean the arrangement like

```
describe("blah blah", () => {
  describe("When blah blah", () => {
    const TX_HASH = '...'

    beforeEach(() => {
      setup()
    })

    it("Should blah blah", () => {
    }
  }
}
```

Reply:
Yes

Reply:
Could you show me some real-world examples? Because most of the libraries I've read follow the same style of the assertion message that I think `it("Should blah blah when/if", () => {}` is more general.

Reply:
This is the spec style of writing tests. I think this article makes a good introduction to the benefits of adopting the `spec` style in a comparison with the traditional `test` style: https://www.bignerdranch.com/blog/why-do-javascript-test-frameworks-use-describe-and-beforeeach/

This is a real-world example (I might be biased since this is a project I participated): https://github.com/hubiinetwork/nahmii-sdk/blob/master/lib/wallet/wallet.spec.js

Reply:
I'm still confused. According to the article and examples, should the code above be changed into the following?

```js
describe("blah blah", () => {
  let txHash

  describe("When tx hash is valid", () => {
    beforeEach(() => {
      txHash = '0x...'
    }
    it("Should blah blah", () => {})
  })

  describe("When tx hash is invalid", () => {
    beforeEach(() => {
      txHash = 'invalid tx hash'
    }
    it("Should blah blah", () => {})
  })
})
```

Reply:
Yes, I think the example you outlined is now much easier to comprehend and to reuse the setups in a spec group. Feel free to let me know if there are still bits unclear.

Reply:
It's clear but not convincing enough. A lot of libraries I've read, e.g. `lodash`, `ramda`, etc. don't have the redundant nesting, even the `jest` is also using the `should blah blah when blah blah` style assertion message. I prefer the general one.

Reply:
And I'm not sure how to express 2 booleans with the `describe(when)` style, should it look like the following?
```js
describe(`when x is true, y is true`, () => {
  beforeEach(() => {
    x = true
    y = true
  })
})

describe(`when x is true, y is false`, () => {
  beforeEach(() => {
    x = true
    y = false
  })
})

describe(`when x is false, y is true`, () => {
  beforeEach(() => {
    x = false
    y = true
  })
})

describe(`when x is false, y is false`, () => {
  beforeEach(() => {
    x = false
    y = false
  })
})
```

Reply:
🤔 Well, I think it’s not a serious problem, I’ll update the style tomorrow. But we need more time to update other tests to keep the style consistent.

Reply:
There are two alternatives to do the same tests in your example.

`jest` or other popular testing frameworks, such as `mocha`, supports nested test suits. Nested test suits make it straightforward to extend the existing tests and for the new tests to reuse the setups in the parent test suit. For example:
```
describe(`when x is true`, () => {
  beforeEach(() => {
    x = true
  })
  describe('when y is true', () => {
    beforeEach(() => {
      y = true
    })
    it()
  })
  describe('when y is false', () => {
    beforeEach(() => {
      y = false
    })
    it()
  })
})
```

The other way is to create dynamic tests using a `for` loop. For example: 
```
[
  [true, true, expected],
  [true, false, expected],
  [false, true, expected],
  [false, false, expected]
].forEach(([_y, _x, expected]) => {
  describe(`when y = ${y} and x = ${x}`, () => {
    let x
    let y
    beforeEach(() => {
      x = _x
      y = _y
    })
    it('should', () => {})
  })
})
```

Reply:
Updated in https://github.com/nervosnetwork/neuron/pull/1733/commits/37d656e18627748973fee2915c2c1bbf4ab74790